Implement GraphFrame and integrate it in VisualShader

Geometror opened the pull request:
This PR introduces `GraphFrame`, a new `GraphElement` designed for organizing node graphs, and integrates it into the VisualShader editor (can be split if desired).
It aims to reimplement the functionality that was lost when comment nodes were removed. While it works a bit differently (very similar to Blender) it should be a lot more predictable and much less problematic compared to the old implementation of comment nodes.

Finally closes godotengine/godot-proposals#5271.

![grafik](https://github.com/godotengine/godot/assets/50084500/03a4ac8d-b917-4304-9b9f-fc63d976329c)

https://github.com/godotengine/godot/assets/50084500/d81c7cf9-6b3d-4d44-b9ae-07f7af88dc43

## Detailed changes:
- Add `GraphFrame`
	- A special graph element to which you can attach other graph elements (nodes and frames). Its rect automatically adapts to the attached elements (enclosing), if autoshrink is enabled. Only attached are moved together with the frame.
	- Can be tinted in an arbitrary color
	- GraphEdit automatically sorts the frames topologically (nested frames are supported)
	- GraphEdit automatically puts frames behind the connection layer
		- This requires to make the connection layer a non-internal child again (reintroduces #85005) since there is no better solution at the moment to keep it between background nodes, e.g. frames, and connectable nodes (just modifying the z_index of each node type would mess up the input and is just confusing)
- VisualShader integration
	- Remove `VisualShaderNodeComment` and replace it with `VisualShaderNodeFrame`
	- Right click context menu: Detach selected nodes from their "parent" frame
	- To attach nodes to a frame simply drag and drop them onto it.
	- Context menu option to enable/disable autoshrink and tint color (as well as choose the tint color)

![grafik](https://github.com/godotengine/godot/assets/50084500/3dc66fa0-b5d3-4d45-930b-814a9df3efc8)

## API
*Note for users of GraphEdit: Integrating GraphFrame in your application is a bit more complicated that the old comment nodes (where you could just check a property and it "worked" out of the box), especially for Undo-Redo, but in return it should work more reliable and be a lot more useful.*

- GraphEdit
	- Methods
		- `attach_graph_element_to_frame(element, frame)` : Attach a node to a frame (params are string names)
		- `detach_graph_element_from_frame(element)` : Detach a node from its frame (param is string name)
		- `get_frame(element)` : Get the frame a node is attached to
	- Signals
		- `frame_size_changed (frame, new_size)` : Emitted when the size of a frame changes
		- `nodes_linked_to_frame_request(elements, frame)` : Emitted when nodes are dropped onto a frame
- GraphElement
	- Signals
		- `resize_end` : Emitted after releasing the mouse button after resizing the node
- GraphFrame
	- Properties
		- `title` : The title of the frame
		- `autoshrink_enabled` : Whether the frame should automatically shrink to fit its contents (attached nodes/frames)
		- `autoshrink_margin` : The margin to add to the frame size on each side when autoshrink is enabled
		- `drag_margin` : The margin around the frame where it can be dragged
		- `tint_color_enabled` : Whether the frame should be tinted in a color
		- `tint_color` : The color to tint the frame in
	- Methods
		- `get_titlebar_hbox` : Get the HBoxContainer of the titlebar (useful for customizing the titlebar)
	- Signals
		- `autoshrink_toggled` : Emitted when the autoshrink property is toggled
- VisualShader
	- Methods
		- `attach_node_to_frame(node, frame)` : Attach a node to a frame
		- `detach_node_from_frame(node)` : Detach a node from its frame

## TODO:
- [x] Theme cache
- [x] Docs
- [x] Address remaining TODO comments

_Production edit: closes godotengine/godot-roadmap#13_

Geometror committed:
Add GraphFrame and integrate it in VisualShader

Mickeon commented on doc/classes/GraphFrame.xml:
```diff
@@ -0,0 +1,69 @@
+			Title of the frame.
+		</member>
+		<member name="title_centered" type="bool" setter="set_title_centered" getter="is_title_centered" default="true">
+			Determines whether the title is aligned to the left/right or centered.
```
Grrrrr- Bark, bark, bark.
```suggestion
			If [code]true[/code] the title is centered. Otherwise, the title is aligned to the left or right.
```

Mickeon commented on doc/classes/VisualShaderNodeFrame.xml:
```diff
@@ -0,0 +1,49 @@
+	<description>
+		A rectangular frame that can be used to group visual shader nodes together to improve organization. 
+		Nodes attached to the frame will move with it when it is dragged and it can automatically resize to enclose all attached nodes.
+		It's title, description and color can be customized.
```
```suggestion
		Its title, description and color can be customized.
```

AThousandShips commented on doc/classes/GraphElement.xml:
```diff
@@ -17,7 +17,7 @@
 		<member name="resizable" type="bool" setter="set_resizable" getter="is_resizable" default="false">
 			If [code]true[/code], the user can resize the GraphElement.
-			[b]Note:[/b] Dragging the handle will only emit the [signal resize_request] signal, the GraphElement needs to be resized manually.
+			[b]Note:[/b] Dragging the handle will only emit the [signal resize_request] and [signal resize_end] signal, the GraphElement needs to be resized manually.
```
```suggestion
			[b]Note:[/b] Dragging the handle will only emit the [signal resize_request] and [signal resize_end] signals, the GraphElement needs to be resized manually.
```

AThousandShips commented on scene/gui/graph_edit.cpp:
```diff
@@ -1926,6 +2250,62 @@ bool GraphEdit::is_valid_connection_type(int p_type, int p_with_type) const {
 
+void GraphEdit::attach_graph_element_to_frame(const StringName &p_graph_element, const StringName &p_parent_frame) {
+	GraphFrame *frame = Object::cast_to<GraphFrame>(get_node(NodePath(p_parent_frame)));
+	if (!frame) {
+		print_line("Parent frame does not exist or is not a GraphFrame.");
+	}
+	ERR_FAIL_COND_MSG(!frame, "Parent frame does not exist or is not a GraphFrame.");
```
```suggestion
	ERR_FAIL_NULL_MSG(frame, "Parent frame does not exist or is not a GraphFrame.");
```

AThousandShips commented on scene/gui/graph_edit.cpp:
```diff
@@ -1926,6 +2250,62 @@ bool GraphEdit::is_valid_connection_type(int p_type, int p_with_type) const {
+
+	GraphFrame *frame = Object::cast_to<GraphFrame>(get_node(NodePath(linked_parent_map[p_graph_element])));
+
+	if (!frame) {
+		print_line("Parent frame does not exist or is not a GraphFrame.");
+	}
+
+	ERR_FAIL_COND_MSG(!frame, "Parent frame does not exist or is not a GraphFrame.");
```
```suggestion
	ERR_FAIL_NULL_MSG(frame, "Parent frame does not exist or is not a GraphFrame.");
```

Geometror commented on doc/classes/GraphFrame.xml:
```diff
@@ -0,0 +1,69 @@
+	<signals>
+		<signal name="autoshrink_toggled">
+			<description>
+				Emitted when the frame's autoshrink state changes.
```
It should (probably) only be emitted if the autoshrink_enabled property changes.

AThousandShips commented on scene/resources/visual_shader.cpp:
```diff
@@ -1129,6 +1141,36 @@ bool VisualShader::is_port_types_compatible(int p_a, int p_b) const {
 
+void VisualShader::attach_node_to_frame(Type p_type, int p_node, int p_frame) {
+	ERR_FAIL_INDEX(p_type, TYPE_MAX);
+	Graph *g = &graph[p_type];
+
+	ERR_FAIL_COND(!g->nodes.has(p_node));
+	ERR_FAIL_COND(p_frame < 0);
```
```suggestion
	ERR_FAIL_COND(p_frame < 0);
	Graph *g = &graph[p_type];

	ERR_FAIL_COND(!g->nodes.has(p_node));
```
I'd say to not delay after lookup for this check

Geometror commented on doc/classes/GraphEdit.xml:
```diff
@@ -395,6 +417,20 @@
+			<param index="0" name="elements" type="Array" />
+			<param index="1" name="frame" type="StringName" />
+			<description>
+				Emitted when [GraphElement]s are dropped onto a [GraphFrame], when they were not previously attached to any other one.
```
Went with 
```
Emitted when one or more [GraphElement]s are dropped onto the [GraphFrame] named [param frame], when they were not previously attached to any other one.
[param elements] is an array of [GraphElement]s to be attached
```

Geometror commented on scene/gui/graph_edit.cpp:
```diff
@@ -29,6 +29,7 @@
 /**************************************************************************/
 
 #include "graph_edit.h"
+
```
This is intetional. When I remove this line, clang-format sorts it and it ends up before `#include "graph_edit.h"` which then results in a compilation error.

AThousandShips replied:
That's very odd! It works everywhere else

Geometror replied:
I just checked it with the other files, same for me there (using clang-format 17.0.6)
But I'm actually confused why it worked everywhere else in the first place since
```
#include "graph_edit.compat.inc"
#include "graph_edit.h"
```
is the lexicographically correct order.

AThousandShips replied:
Unsure, some quirk possibly, but same with 17.0.6 and it works fine with this file specifically

Geometror replied:
Mmh, I still have no idea why this doesn't work for me, but I changed it manually for now.

AThousandShips replied:
Tested with 16.0.4 as well and it works as above, so probably some strange bug there, worth a look later

Geometror commented:
@Yagich 
1. Yes, it should. You can test it with the VS expression node which is resizable.
2. I assume you mean the connectable nodes (`GraphNode`s) with "slot children". Then yes, you would need to check whether the children are of type `GraphElement` or `GraphNode` (depending on whether you want the frame nodes too). Just stripping the first child from the children array would only work if you don't have any GraphFrames in there. What we could do however, is expose a method `get_connection_layer_index` so that you could just use that instead of type checking (but then again type checking is a bit safer and not that expensive).

Yagich commented:
I've tested the PR and overall it works great! I've only been able to find one issue: if you try to attach a frame to itself, the application will hang. It's not unfixable in user code, but I wonder if it should just be a no-op if `element` and `frame` are the same when calling `attach_graph_element_to_frame()`.

Yagich commented:
> I assume you mean the connectable nodes (`GraphNode`s) with "slot children"

Sorry, I think I got confused what issue this refers to when writing the original comment. What I meant was that since the connection layer is now not internal, every `get_children()` call on `GraphEdit` to get elements now has to slice off the first child. I suppose there really is no way around it now.

Yagich commented on scene/gui/graph_edit.cpp:
```diff
@@ -1926,6 +2249,55 @@ bool GraphEdit::is_valid_connection_type(int p_type, int p_with_type) const {
+	return Object::cast_to<GraphFrame>(parent);
+}
+
+TypedArray<StringName> GraphEdit::get_attached_noded_of_frame(const StringName &p_graph_frame) {
```
typo: `get_attached_noded_of_frame` -> `get_attached_nodes_of_frame`

Geometror replied:
I also haven't exposed this method by accident. Fixed.

Geometror commented:
@Yagich Thanks for testing. I made both `attach_graph_element_to_frame()` and `detach_graph_element_from_frame()` more robust.

Mickeon commented on doc/classes/GraphFrame.xml:
```diff
@@ -0,0 +1,69 @@
+		<member name="title" type="String" setter="set_title" getter="get_title" default="&quot;&quot;">
+			Title of the frame.
+		</member>
+		<member name="title_centered" type="bool" setter="set_title_centered" getter="is_title_centered" default="true">
+			If [code]true[/code] the title is centered. Otherwise, the title is aligned to the left or right.
+		</member>
```
In https://github.com/godotengine/godot/pull/88014#discussion_r1481433307 I did agree with you and said that this property should probably not exist. If users need it they will ask for it.

Geometror replied:
Removed it.

Geometror commented:
Rebased and adjusted the style a bit after #85017 has been merged.
![image](https://github.com/godotengine/godot/assets/50084500/10cdbd1e-ae05-49ee-bc14-114da430752e)
![image](https://github.com/godotengine/godot/assets/50084500/311d27c0-8a6d-4e53-b2bf-6d00fa59d546)

akien-mga commented:
Tested briefly, I'm not familiar with Visual Shader so I'm just trying some random stuff.

I created a Visual Shader, attached to a Sprite2D (icon.svg).

Pressed "Add Node...", dragged and dropped ColorConstant to the graph, I get those two errors:
```
ERROR: Condition "!get_viewport()->gui_is_dragging()" is true.
   at: set_drag_preview (./scene/gui/control.cpp:1965)
ERROR: Graph element is not attached to any frame.
   at: detach_graph_element_from_frame (./scene/gui/graph_edit.cpp:2269)
```

The frame error gets repeated every time I do an action, e.g. toggling visibility of the ColorConstant, or connecting it to Output > Color.

I then found how to add a frame via right click/Add Node, which auto-shrunk to tiny. I had to right click it to disable auto-shrink, so I could extend it to encompass three nodes I had.

Then it seemed only the Output node within the frame was actually attached to the frame. It's not clear to me how to attach nodes to a frame. I did a bunch of random single or double clicking on nodes, moving them a bit within the frame, and they somehow ended up being attached, though there's visible feedback about it.

The end result works, but as a first time user this wasn't very intuitive.
![image](https://github.com/godotengine/godot/assets/4701338/f50380c4-8c13-479f-b24a-fa2857fe60fc)

The frame description also can easily overlap with nodes:
![image](https://github.com/godotengine/godot/assets/4701338/955cc228-af4a-46d2-8a0e-692b72b18c37)

This was a really brief test on laptop with touchpad, so hard mode :P
Hope it's useful anyway :)

BlueCube3310 commented:
Bug report: Duplicating a frame with elements inside of it will cause the elements inside to be attached to both the original and the duplicate.

Geometror commented:
@akien-mga @BlueCube3310 Thanks for testing! 
All the described (technical) issues should be solved now.

Regarding the usability problems:
I will experiment a bit with a cursor hint (when dragging a unattached node over a frame), a hint text for empty frames and maybe a small visual indicator that a node is attached (although that might be too much and if not likely part of a subsequent PR due to limited time).

I'm thinking about removing the description feature and instead add a true "comment" node which is basically a dragable label with a full transparent background (which would behave the same way as normal node inside a frame, so you could place it freely). IMO this would be much more intuitive.

> BTW, this is likely not for this PR, but do you plan to add support for duplicating frames with their contents, and/or saving frames with their contents as a reusable component? I'm not familiar with Visual Shader but this sounds like it could be quite useful.

Yes, I plan to do something similar (https://github.com/godotengine/godot-proposals/issues/7730) 
Although that is completely separate from graph frames, which are purely for visual organization.

Chaosus commented:
> Pressed "Add Node...", dragged and dropped ColorConstant to the graph, I get those two errors:
> 
> ERROR: Condition "!get_viewport()->gui_is_dragging()" is true.
>    at: set_drag_preview (./scene/gui/control.cpp:1965)

@akien-mga The first message is related to old bug: https://github.com/godotengine/godot/issues/59622, should be fixed by https://github.com/godotengine/godot/pull/67531

Geometror commented:
Small update: Refactoring on the visual shader side and a hint for empty frames:
![frame drop](https://github.com/godotengine/godot/assets/50084500/199dc126-69ec-4832-8393-e2089cac4767)

KoBeWi commented:
You can't undo adding a frame and trying to delete one results in a crash:
```
CrashHandlerException: Program crashed
Engine version: Godot Engine v4.3.dev.custom_build (5d08c2631cadf29d80ca23c7d537e03c3e5edcc5)
Dumping the backtrace. Please include this when reporting the bug to the project developer.
[0] VisualShaderNode::is_deletable (C:\godot_source\scene\resources\visual_shader.cpp:352)
[1] VisualShaderNode::is_deletable (C:\godot_source\scene\resources\visual_shader.cpp:352)
[2] VisualShaderEditor::_delete_nodes_request (C:\godot_source\editor\plugins\visual_shader_editor_plugin.cpp:4346)
[3] call_with_variant_args_helper<VisualShaderEditor,TypedArray<StringName> const &,0> (C:\godot_source\core\variant\binder_common.h:304)
[4] call_with_variant_args<VisualShaderEditor,TypedArray<StringName> const &> (C:\godot_source\core\variant\binder_common.h:419)
[5] CallableCustomMethodPointer<VisualShaderEditor,TypedArray<StringName> const &>::call (C:\godot_source\core\object\callable_method_pointer.h:104)
[6] Callable::callp (C:\godot_source\core\variant\callable.cpp:58)
[7] Object::emit_signalp (C:\godot_source\core\object\object.cpp:1202)
[8] Node::emit_signalp (C:\godot_source\scene\main\node.cpp:3904)
[9] Object::emit_signal<TypedArray<StringName> > (C:\godot_source\core\object\object.h:933)
[10] GraphEdit::gui_input (C:\godot_source\scene\gui\graph_edit.cpp:1972)
[11] Control::_call_gui_input (C:\godot_source\scene\gui\control.cpp:1800)
[12] Viewport::_gui_input_event (C:\godot_source\scene\main\viewport.cpp:2178)
[13] Viewport::push_input (C:\godot_source\scene\main\viewport.cpp:3229)
[14] Window::_window_input (C:\godot_source\scene\main\window.cpp:1627)
[15] call_with_variant_args_helper<Window,Ref<InputEvent> const &,0> (C:\godot_source\core\variant\binder_common.h:304)
[16] call_with_variant_args<Window,Ref<InputEvent> const &> (C:\godot_source\core\variant\binder_common.h:419)
[17] CallableCustomMethodPointer<Window,Ref<InputEvent> const &>::call (C:\godot_source\core\object\callable_method_pointer.h:104)
[18] Callable::callp (C:\godot_source\core\variant\callable.cpp:58)
[19] Callable::call<Ref<InputEvent> > (C:\godot_source\core\variant\variant.h:864)
[20] DisplayServerWindows::_dispatch_input_event (C:\godot_source\platform\windows\display_server_windows.cpp:3450)
[21] DisplayServerWindows::_dispatch_input_events (C:\godot_source\platform\windows\display_server_windows.cpp:3420)
[22] Input::_parse_input_event_impl (C:\godot_source\core\input\input.cpp:773)
[23] Input::flush_buffered_events (C:\godot_source\core\input\input.cpp:1044)
[24] DisplayServerWindows::process_events (C:\godot_source\platform\windows\display_server_windows.cpp:2969)
[25] OS_Windows::run (C:\godot_source\platform\windows\os_windows.cpp:1476)
[26] widechar_main (C:\godot_source\platform\windows\godot_windows.cpp:181)
[27] _main (C:\godot_source\platform\windows\godot_windows.cpp:206)
[28] main (C:\godot_source\platform\windows\godot_windows.cpp:220)
[29] WinMain (C:\godot_source\platform\windows\godot_windows.cpp:234)
[30] __scrt_common_main_seh (D:\a\_work\1\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl:288)
[31] <couldn't map PC to fn name>
-- END OF BACKTRACE --
```

KoBeWi commented on doc/classes/GraphFrame.xml:
```diff
@@ -0,0 +1,66 @@
+			The margin around the frame (horizontal, vertical) that is used to calculate the size of the frame when [member autoshrink_enabled] is [code]true[/code].
+		</member>
+		<member name="drag_margin" type="int" setter="set_drag_margin" getter="get_drag_margin" default="16">
+			The margin around the frame that can be used to drag the frame.
```
This seems to be a margin *inside* the frame. It should be clarified.

KoBeWi commented:
The crashes are now fixed, but I still have the problem with detaching. I'm on Windows if that matters.

> What was your setup? Sounds like you added GraphNodes as children to a GraphFrame, which is not how it works (due to layering issues it's not possible this way -> connection lines). Nodes have to be attached to a frame using a GraphEdit method.

Yep, I just used it wrongly.

> I'm not sure what to do with the resizable property. It makes not much sense for frames, unless we would remove the autoshrink prop and use it as a replacement.

You can hide it at least, using `_validate_property()`.

KoBeWi commented on doc/classes/GraphEdit.xml:
```diff
@@ -179,6 +201,13 @@
 				Returns an [Array] containing the list of connections that intersect with the given [Rect2]. A connection consists in a structure of the form [code]{ from_port: 0, from_node: "GraphNode name 0", to_port: 1, to_node: "GraphNode name 1" }[/code].
 			</description>
 		</method>
+		<method name="get_frame">
```
```suggestion
		<method name="get_element_frame">
```
That would be more accurate to what it does.

Geometror replied:
Oh, yes, I wanted to rename this but forgot :)

KoBeWi commented:
Changing `autoshrink_margin` with a node attached to a frame will not automatically redraw the frame.

https://github.com/godotengine/godot/assets/2223172/285e1e71-0fca-48e0-95c9-eb0a9a3d65ee

KoBeWi commented on scene/gui/graph_frame.h:
```diff
@@ -0,0 +1,108 @@
+	GDCLASS(GraphFrame, GraphElement);
+
+	struct _MinSizeCache {
+		int min_size;
+		bool will_stretch;
+		int final_size;
```
```suggestion
		int min_size = 0;
		bool will_stretch = 0;
		int final_size = 0;
```

KoBeWi commented on scene/gui/graph_frame.h:
```diff
@@ -0,0 +1,108 @@
+class GraphFrame : public GraphElement {
+	GDCLASS(GraphFrame, GraphElement);
+
+	struct _MinSizeCache {
```
Why is the name prefixed?

Geometror replied:
Was copied from box_container.cpp, I thought this was a convention for private, single-use structs.

KoBeWi replied:
BoxContainer code is super old and was probably never updated.
Though I don't recall any other case of such struct.

Geometror replied:
I just did a quick search and it looks like prefixing structs defined in cpp files (or private inner structs) with _ seems to be common practice in our code base, even in newer code.

KoBeWi replied:
Huh. Then it's used inconsistently 🤔

KoBeWi commented on scene/gui/graph_frame.cpp:
```diff
@@ -0,0 +1,351 @@
+	Ref<InputEventMouseMotion> mm = p_ev;
+
+	// Only resize if the frame is not auto-resizing based on linked nodes.
+	if (resizing && mm.is_valid() && !autoshrink_enabled) {
```
```suggestion
	if (resizing && !autoshrink_enabled && mm.is_valid()) {
```
Though autoshrink check shouldn't be here I think, because `resizable` depends on it anyway. In unlikely situation where autoshrink changes during resizing, it will never be released.

KoBeWi commented on scene/gui/graph_edit.cpp:
```diff
@@ -2374,7 +2761,7 @@ GraphEdit::GraphEdit() {
 
 	connections_layer = memnew(Control);
-	add_child(connections_layer, false, INTERNAL_MODE_FRONT);
+	add_child(connections_layer, false);
```
Why did you remove internal mode? It makes the connection layer easily accessible by the user.

Geometror replied:
I tried several approaches (e.g. using the z-index plus some workarounds) and this was actually the only viable solution. I don't see how can possibly fix this without a major comfpat breakage later or introducing internal nodes at arbitrary positions.

KoBeWi commented on scene/gui/graph_edit.cpp:
```diff
@@ -442,11 +442,42 @@ void GraphEdit::_update_scroll() {
+	List<GraphFrame *> attached_nodes_to_move;
+	attached_nodes_to_move.push_back(frame);
+
+	while (attached_nodes_to_move.size()) {
```
```suggestion
	while (!attached_nodes_to_move.is_empty()) {
```

KoBeWi commented on scene/gui/graph_edit.cpp:
```diff
@@ -442,11 +442,42 @@ void GraphEdit::_update_scroll() {
+			continue;
+		}
+
+		for (StringName attached_node_name : frame_attached_nodes.get(attached_frame->get_name())) {
```
```suggestion
		for (const StringName &attached_node_name : frame_attached_nodes.get(attached_frame->get_name())) {
```

KoBeWi commented on scene/gui/graph_edit.h:
```diff
@@ -304,12 +313,15 @@ class GraphEdit : public Control {
 	void _graph_element_deselected(Node *p_node);
-	void _graph_element_moved_to_front(Node *p_node);
-	void _graph_element_resized(Vector2 p_new_minsize, Node *p_node);
+	void _graph_element_resize_request(Vector2 p_new_minsize, Node *p_node);
+	void _graph_frame_shrink_toggled(Vector2 p_new_minsize, GraphFrame *p_node);
```
```suggestion
	void _graph_element_resize_request(const Vector2 &p_new_minsize, Node *p_node);
	void _graph_frame_shrink_toggled(const Vector2 &p_new_minsize, GraphFrame *p_node);
```

KoBeWi commented on scene/gui/graph_edit.cpp:
```diff
@@ -520,10 +602,24 @@ void GraphEdit::add_child_notify(Node *p_child) {
+		if (graph_frame) {
+			background_nodes_separator_idx++;
+
+			call_deferred("move_child", graph_frame, 0);
+			call_deferred("_reorder_frame_nodes", graph_frame->get_name());
+			call_deferred("move_child", connections_layer, background_nodes_separator_idx);
```
```suggestion
			callable_mp((Node *)this, &Node::move_child).call_deferred(graph_frame, 0);
			callable_mp((Node *)this, &Node::move_child).call_deferred(connections_layer, background_nodes_separator_idx);
```
`_reorder_frame_nodes` does not exist.

KoBeWi commented on scene/gui/graph_edit.cpp:
```diff
@@ -565,8 +661,36 @@ void GraphEdit::remove_child_notify(Node *p_child) {
+					frame_attached_nodes.get(parent_frame->get_name()).erase(graph_element->get_name());
+				}
+				linked_parent_map.erase(graph_element->get_name());
+				call_deferred("_reorder_frame_nodes", parent_frame->get_name());
```
```suggestion
```

AThousandShips commented on scene/resources/visual_shader.cpp:
```diff
@@ -4160,66 +4208,120 @@ VisualShaderNodeResizableBase::VisualShaderNodeResizableBase() {
+	ClassDB::bind_method(D_METHOD("set_tint_color", "p_color"), &VisualShaderNodeFrame::set_tint_color);
+	ClassDB::bind_method(D_METHOD("get_tint_color"), &VisualShaderNodeFrame::get_tint_color);
+
+	ClassDB::bind_method(D_METHOD("set_autoshrink_enabled", "p_enable"), &VisualShaderNodeFrame::set_autoshrink_enabled);
```
```suggestion
	ClassDB::bind_method(D_METHOD("set_autoshrink_enabled", "enable"), &VisualShaderNodeFrame::set_autoshrink_enabled);
```

KoBeWi commented on scene/gui/graph_edit.cpp:
```diff
@@ -442,11 +442,42 @@ void GraphEdit::_update_scroll() {
+		attached_nodes_to_move.pop_front();
+
+		// Move the frame to the front of the background node index range.
+		attached_frame->get_parent()->call_deferred("move_child", attached_frame, background_nodes_separator_idx - 1);
```
```suggestion
		callable_mp((Node *)attached_frame, &Node::move_child).call_deferred(attached_frame, background_nodes_separator_idx - 1);
```

AThousandShips commented on scene/resources/visual_shader.cpp:
```diff
@@ -4160,66 +4208,120 @@ VisualShaderNodeResizableBase::VisualShaderNodeResizableBase() {
 
-	ClassDB::bind_method(D_METHOD("set_description", "description"), &VisualShaderNodeComment::set_description);
-	ClassDB::bind_method(D_METHOD("get_description"), &VisualShaderNodeComment::get_description);
+	ClassDB::bind_method(D_METHOD("add_attached_node", "p_node"), &VisualShaderNodeFrame::add_attached_node);
```
```suggestion
	ClassDB::bind_method(D_METHOD("add_attached_node", "node"), &VisualShaderNodeFrame::add_attached_node);
```

AThousandShips commented on scene/resources/visual_shader.cpp:
```diff
@@ -4160,66 +4208,120 @@ VisualShaderNodeResizableBase::VisualShaderNodeResizableBase() {
-	ClassDB::bind_method(D_METHOD("set_description", "description"), &VisualShaderNodeComment::set_description);
-	ClassDB::bind_method(D_METHOD("get_description"), &VisualShaderNodeComment::get_description);
+	ClassDB::bind_method(D_METHOD("add_attached_node", "p_node"), &VisualShaderNodeFrame::add_attached_node);
+	ClassDB::bind_method(D_METHOD("remove_attached_node", "p_node"), &VisualShaderNodeFrame::remove_attached_node);
```
```suggestion
	ClassDB::bind_method(D_METHOD("remove_attached_node", "node"), &VisualShaderNodeFrame::remove_attached_node);
```

AThousandShips commented on scene/resources/visual_shader.cpp:
```diff
@@ -4160,66 +4208,120 @@ VisualShaderNodeResizableBase::VisualShaderNodeResizableBase() {
+	ClassDB::bind_method(D_METHOD("set_tint_color_enabled", "p_enable"), &VisualShaderNodeFrame::set_tint_color_enabled);
+	ClassDB::bind_method(D_METHOD("is_tint_color_enabled"), &VisualShaderNodeFrame::is_tint_color_enabled);
+
+	ClassDB::bind_method(D_METHOD("set_tint_color", "p_color"), &VisualShaderNodeFrame::set_tint_color);
```
```suggestion
	ClassDB::bind_method(D_METHOD("set_tint_color", "color"), &VisualShaderNodeFrame::set_tint_color);
```

KoBeWi commented on scene/gui/graph_edit.cpp:
```diff
@@ -565,8 +661,36 @@ void GraphEdit::remove_child_notify(Node *p_child) {
+		}
+
+		if (frame_attached_nodes.has(graph_element->get_name())) {
+			for (String attached_node_name : frame_attached_nodes.get(graph_element->get_name())) {
```
```suggestion
			for (const StringName &attached_node_name : frame_attached_nodes.get(graph_element->get_name())) {
```

AThousandShips commented on scene/gui/graph_frame.h:
```diff
@@ -0,0 +1,108 @@
+	void set_autoshrink_enabled(bool p_enable);
+	bool is_autoshrink_enabled() const;
+
+	void set_autoshrink_margin(const int &p_margin);
```
```suggestion
	void set_autoshrink_margin(int p_margin);
```

AThousandShips commented on scene/gui/graph_frame.cpp:
```diff
@@ -0,0 +1,351 @@
+	ClassDB::bind_method(D_METHOD("set_tint_color_enabled", "p_enable"), &GraphFrame::set_tint_color_enabled);
+	ClassDB::bind_method(D_METHOD("is_tint_color_enabled"), &GraphFrame::is_tint_color_enabled);
+
+	ClassDB::bind_method(D_METHOD("set_tint_color", "p_color"), &GraphFrame::set_tint_color);
```
```suggestion
	ClassDB::bind_method(D_METHOD("set_tint_color", "color"), &GraphFrame::set_tint_color);
```

AThousandShips commented on scene/gui/graph_frame.cpp:
```diff
@@ -0,0 +1,351 @@
+			Rect2 titlebar_rect(Point2(), titlebar_hbox->get_size() + sb_titlebar->get_minimum_size());
+			Size2 body_size = get_size();
+			body_size.y -= titlebar_rect.size.height;
+			Rect2 body_rect(0, titlebar_rect.size.height, body_size.width, body_size.height);
```
```suggestion
			Rect2 body_rect(Point2(0, titlebar_rect.size.height), body_size);
```

KoBeWi commented on scene/gui/graph_edit.cpp:
```diff
@@ -628,6 +753,118 @@ void GraphEdit::_notification(int p_what) {
+
+	int autoshrink_margin = p_frame->get_autoshrink_margin();
+
+	for (StringName attached_node_name : frame_attached_nodes.get(p_frame->get_name())) {
```
```suggestion
	for (const StringName &attached_node_name : frame_attached_nodes.get(p_frame->get_name())) {
```

KoBeWi commented on scene/gui/graph_edit.cpp:
```diff
@@ -628,6 +753,118 @@ void GraphEdit::_notification(int p_what) {
+		return;
+	}
+
+	for (StringName attached_node_name : frame_attached_nodes.get(p_frame->get_name())) {
```
```suggestion
	for (const StringName &attached_node_name : frame_attached_nodes.get(p_frame->get_name())) {
```

KoBeWi commented on scene/gui/graph_edit.cpp:
```diff
@@ -628,6 +753,118 @@ void GraphEdit::_notification(int p_what) {
+		return;
+	}
+
+	for (StringName attached_node_name : frame_attached_nodes.get(p_frame->get_name())) {
```
```suggestion
	for (const StringName &attached_node_name : frame_attached_nodes.get(p_frame->get_name())) {
```

KoBeWi commented on editor/plugins/visual_shader_editor_plugin.cpp:
```diff
@@ -282,6 +290,98 @@ void VisualShaderGraphPlugin::set_expression(VisualShader::Type p_type, int p_no
+		return;
+	}
+
+	StringName node_name = itos(p_node_id);
```
```suggestion
	const StringName node_name = itos(p_node_id);
```

KoBeWi commented on editor/plugins/visual_shader_editor_plugin.cpp:
```diff
@@ -1154,8 +1327,14 @@ void VisualShaderGraphPlugin::add_node(VisualShader::Type p_type, int p_id, bool
+			return;
+		}
+
+		StringName graph_element_name = links[p_id].graph_element->get_name();
```
This seems unused.

KoBeWi commented on editor/plugins/visual_shader_editor_plugin.cpp:
```diff
@@ -2532,38 +2723,30 @@ void VisualShaderEditor::_set_node_size(int p_type, int p_node, const Vector2 &p
 	EditorUndoRedoManager *undo_redo = EditorUndoRedoManager::get_singleton();
-	undo_redo->create_action(TTR("Resize VisualShader Node"), UndoRedo::MERGE_ENDS);
-	undo_redo->add_do_method(this, "_set_node_size", p_type, p_node, new_size);
+	undo_redo->create_action(TTR("Resize VisualShader Node"), UndoRedo::MERGE_DISABLE);
```
```suggestion
	undo_redo->create_action(TTR("Resize VisualShader Node"));
```
It's default.

KoBeWi commented on editor/plugins/visual_shader_editor_plugin.cpp:
```diff
@@ -2587,95 +2770,154 @@ void VisualShaderEditor::_preview_select_port(int p_node, int p_port) {
-	ERR_FAIL_COND(!comment_title_change_popup->has_meta("id"));
-	int node_id = (int)comment_title_change_popup->get_meta("id");
+void VisualShaderEditor::_frame_title_popup_hide() {
+	ERR_FAIL_COND(!frame_title_change_popup->has_meta("id"));
+	int node_id = (int)frame_title_change_popup->get_meta("id");
```
```suggestion
	int node_id = (int)frame_title_change_popup->get_meta("id", -1);
	ERR_FAIL_COND(node_id == -1);
```
Slightly better (assuming -1 is invalid ID).

KoBeWi reviewed (approved):
Looks fine now (except the `autoshrink_margin` issue, but it's less important).
I left some style comments, otherwise it's good to go.

AThousandShips commented on scene/gui/graph_edit.h:
```diff
@@ -332,6 +343,10 @@ class GraphEdit : public Control {
 
+	Rect2 _compute_shrinked_frame_rect(const GraphFrame *p_frame);
+	void _set_drag_frame_contained_nodes(GraphFrame *p_frame, bool p_drag);
+	void _set_position_of_frame_contained_nodes(GraphFrame *p_frame, Vector2 p_pos);
```
```suggestion
	void _set_position_of_frame_contained_nodes(GraphFrame *p_frame, const Vector2 &p_pos);
```

AThousandShips commented on scene/gui/graph_frame.cpp:
```diff
@@ -0,0 +1,351 @@
+	return autoshrink_enabled;
+}
+
+void GraphFrame::set_autoshrink_margin(const int &p_margin) {
```
```suggestion
void GraphFrame::set_autoshrink_margin(int p_margin) {
```

AThousandShips commented on scene/gui/graph_edit.cpp:
```diff
@@ -463,11 +494,42 @@ void GraphEdit::_graph_element_deselected(Node *p_node) {
 }
 
-void GraphEdit::_graph_element_resized(Vector2 p_new_minsize, Node *p_node) {
+void GraphEdit::_graph_element_resize_request(Vector2 p_new_minsize, Node *p_node) {
```
```suggestion
void GraphEdit::_graph_element_resize_request(const Vector2 &p_new_minsize, Node *p_node) {
```

AThousandShips commented on scene/gui/graph_edit.cpp:
```diff
@@ -463,11 +494,42 @@ void GraphEdit::_graph_element_deselected(Node *p_node) {
+	}
+}
+
+void GraphEdit::_graph_frame_shrink_toggled(Vector2 p_new_minsize, GraphFrame *p_frame) {
```
```suggestion
void GraphEdit::_graph_frame_shrink_toggled(const Vector2 &p_new_minsize, GraphFrame *p_frame) {
```

AThousandShips commented on scene/gui/graph_edit.cpp:
```diff
@@ -628,6 +751,118 @@ void GraphEdit::_notification(int p_what) {
+
+	// It's sufficient to check only one value here.
+	if (min_point.x == FLT_MAX) {
+		return Rect2(p_frame->get_position_offset(), Size2(0, 0));
```
```suggestion
		return Rect2(p_frame->get_position_offset(), Size2());
```

AThousandShips commented on scene/gui/graph_edit.cpp:
```diff
@@ -628,6 +751,118 @@ void GraphEdit::_notification(int p_what) {
+	}
+}
+
+void GraphEdit::_set_position_of_frame_contained_nodes(GraphFrame *p_frame, Vector2 p_drag_accum) {
```
```suggestion
void GraphEdit::_set_position_of_frame_contained_nodes(GraphFrame *p_frame, const Vector2 &p_pos) {
```

lyuma commented:
@Geometror would it be possible to resolve the conflict with extension_api_validation / squash?

(I believe the conflict was caused by an urgent bugfix with Node::replace_by in #89992)

akien-mga commented:
Thanks!

Zylann commented on scene/gui/graph_frame.cpp:
```diff
@@ -0,0 +1,357 @@
+					sb_panel_flat->set_border_color(selected ? original_border_color : tint_color.lightened(0.3));
+					draw_style_box(sb_panel_flat, body_rect);
+				} else if (sb_panel_texture.is_valid()) {
+					sb_panel_texture = sb_panel_flat->duplicate();
```
This code might crash, as highlighted by its failure to compile on my Linux CI: https://github.com/Zylann/godot_voxel/actions/runs/8560927646/job/23461023064

akien-mga replied:
Reading the code, I think the two identifiers might be swapped bg mistake?

lyuma commented:
>Remove VisualShaderNodeComment and replace it with VisualShaderNodeFrame

@Geometror Removing this part of our API is causing problems with some GDExtension plugins: it wasn't marked experimental, so this breaks all rust plugins since they hard-link to the whole API. See #90303

I think the easiest way to solve this is to add a stub `class VisualShaderNodeComment : public VisualShaderNodeFrame` so it will work just like the new `VisualShaderNodeFrame`. If there is a way to register the GDCLASS without it showing up in the resource picker, that would be best.

Even though it extends from `VisualShaderNodeFrame`, it might also be necessary to implement `bind_compatibility_methods()` for compat versions of set/get_title and set/get_description that call the real versions of those methods.

eobet commented:
It would be easier to find this if when you search for "comment" it would also bring up the frame...
